### PR TITLE
add -fmerge-all-constants flag to LINKFLAGS

### PR DESCRIPTION
**Pull request by breed**
-fmerge-all-constants doesn't fully work if it isn't also included
as part of the link step. This change will add -fmerge-all-constants
to both CCFLAGS and LINKFLAGS if it is specified as a build flag.

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [X] I have updated the appropriate documentation

### CODE REVIEW

**Comment by breed** [comment]
the appveyor error looks like some sort of transient package download problem. is there an easy way to kick it so that it tries again?

**Comment by bdbaddog** [comment]
> the appveyor error looks like some sort of transient package download problem. is there an easy way to kick it so that it tries again?

Done. Running now.

**Comment by bdbaddog** [comment]
You'll need to update: src/engine/SCons/Environment.xml search for "-mno-cygwin" and you should find the section/syntax needed to be updated.

**Comment by mwichmann** [comment]
Looks like `-fopenmp` is not in the current docs, but recognized by `ParseFlags`.

@breed - you do know this only works for `env.MergeFlags` and `env.ParseFlags` calls, right? Else you still have to add it both places by hand. Just checking it doesn't end up a surprise...

**Comment by breed** [comment]
> You'll need to update: src/engine/SCons/Environment.xml search for "-mno-cygwin" and you should find the section/syntax needed to be updated.

sorry ultra n00b question: when i update Environment.xml do i need to regenerated and check in the generated doc?

also, i needed to make the indentation bigger for the example commands list. is there a particular alignment for that in the docs?

**Comment by breed** [comment]
> Looks like `-fopenmp` is not in the current docs, but recognized by `ParseFlags`.
> 
> @breed - you do know this only works for `env.MergeFlags` and `env.ParseFlags` calls, right? Else you still have to add it both places by hand. Just checking it doesn't end up a surprise...

i'm using this with the toolchain for platformio. i believe we are only calling there those functions. this change does result in the functionality i need. (my environment is incredibly resource constrained.)

**Comment by acmorrow** [comment]
I wanted to share the following. My feeling is that if you are using the compiler to drive the link step, you more or less always want to re-iterate the compiler flags on the link line. Our current SCons setup is littered with instances of:

```
if (some_condition_requring_ffoo):
   env.AppendUnique(CCFLAGS=['-ffoo'], LINKFLAGS=['-ffoo'])
```

As an experiment, I came up with the following, though I haven't yet deployed it:


```
    from SCons.Tool.cxx import iscplusplus

    def smartlink_flags(source, target, env, for_signature):
        if iscplusplus(source):
            return '$CXXFLAGS $CCFLAGS'
        return '$CFLAGS $CCFLAGS'
    env['SMARTLINKFLAGS'] = smartlink_flags

    def smartshlink_flags(source, target, env, for_signature):
        if iscplusplus(source):
            return '$SHCXXFLAGS $SHCCFLAGS'
        return '$SHCFLAGS $SHCCFLAGS'
    env['SMARTSHLINKFLAGS'] = smartshlink_flags

    env.PrependUnique(
        LINKFLAGS=['$SMARTLINKFLAGS'],
        SHLINKFLAGS=['$SMARTSHLINKFLAGS'],
    )

```

It hasn't got much testing yet and surely can be improved, but it seemed somewhat relevant to this discussion.

**Comment by mwichmann** [comment]
> > You'll need to update: src/engine/SCons/Environment.xml search for "-mno-cygwin" and you should find the section/syntax needed to be updated.
> 
> sorry ultra n00b question: when i update Environment.xml do i need to regenerated and check in the generated doc?
> 
> also, i needed to make the indentation bigger for the example commands list. is there a particular alignment for that in the docs?

1. Bill usually regens docs when getting ready to make a release so it's not required.
2. The <example_commands> markup is scons-specific and should just come out right... but I'm not even sure it should be used here: it's not example code; examples are extracted and run by the doc build process to capture output)... that bit of musing is not your problem.
3. your unit tests are failing because of ordering of args, which is a little fussy. I'll add a separate comment at the point.

File: src/engine/SCons/EnvironmentTests.py [hunk 1]
```diff
@@ -2025,6 +2027,7 @@ def __call__(self, command):
                                  "-F fwd3 " + \
                                  "-pthread " + \
                                  "-mno-cygwin -mwindows " + \
+                                 "-fmerge-all-constants " + \
```
Comment (mwichmann) [inline src/engine/SCons/EnvironmentTests.py:2029]: here you add the flag after `-mno-cygwin`

File: src/engine/SCons/EnvironmentTests.py [hunk 2]
```diff
@@ -2035,7 +2038,7 @@ def __call__(self, command):
             assert save_command == ['fake command'], save_command
             assert env['ASFLAGS'] == ['assembler', '-as'], env['ASFLAGS']
             assert env['CCFLAGS'] == ['', '-X', '-Wa,-as',
-                                      '-pthread', '-mno-cygwin',
+                                      '-pthread', '-fmerge-all-constants', '-mno-cygwin',
```
Comment (mwichmann) [inline src/engine/SCons/EnvironmentTests.py:2041]: ...and here you expect it to be seen before `-mno-cygwin`

**Comment by bdbaddog** [comment]
If you can wrap this up by the weekend, should be able to include in 3.1.1
If you have spare cycle to add the ```-fopenmp``` to the docs that'd be very welcome as well.

File: src/engine/SCons/EnvironmentTests.py (hunk 2 above)
Comment (breed) [reply src/engine/SCons/EnvironmentTests.py:2041]: ugh! my bad! i didn't run the tests! (and i forgot that i could easily commit this suggestion :'( n00b...

**Comment by breed** [comment]
> 
> 
> If you can wrap this up by the weekend, should be able to include in 3.1.1
> If you have spare cycle to add the `-fopenmp` to the docs that'd be very welcome as well.

it appears that openmp was added years ago. see commit 3d9b98eac
just to make sure that we are on the same page: when you say add to the docs, you mean just adding a line to Environment.xml? i can totally do that. should i do it in this change?

**Comment by bdbaddog** [comment]
@breed  - yes. please add in Environment.xml

**Comment by bdbaddog** [comment]
@breed   - Thanks! Merged.

**Comment by breed** [comment]
thanks for your help @bdbaddog and @mwichmann !

**Pull request merged**